src/som/interpreter/nodes/nary/RedirectReadNode.java

```diff
@@ -0,0 +1,29 @@
+package som.interpreter.nodes.nary;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.source.SourceSection;
+
+import som.compiler.Variable;
+import som.interpreter.nodes.PreevaluatedExpression;
+
+public class RedirectReadNode extends ExprWithTagsNode
+    implements PreevaluatedExpression {
+
+  final VirtualFrame frame;
```

Comment:
First rule of Graal/Truffle: you can never ever save a reference to a virtual frame.
The `VirtualFrame` is supposed to never escape, because it is never supposed to be allocated as an object.
Furthermore, this is an AST node here, an AST node is normally independent of a specific activation.
Please see [`determineContext()`] (https://github.com/smarr/SOMns/blob/master/src/som/interpreter/nodes/ContextualNode.java#L51) and its uses to look up the relevant lexical frame for a variable in the lexical context at run time.